src/Forms/SingleSelectField.php

```diff
@@ -50,6 +47,23 @@ public function getSchemaDataDefaults()
         return $data;
     }
 
+    public function setSubmittedValue($value, $data = null)
+    {
+        if (is_numeric($value) && is_string($value) && $this->getAllSourceValuesAreIntType()) {
+            $value = (int) $value;
+        }
+        parent::setSubmittedValue($value, $data);
+    }
+
```

Comment:
Yup, can confirm this now gives a "Must be a string" validation error message which a) makes no sense for a user who's just selecting an item from a dropdown and b) didn't used to be a problem. int or string keys should work perfectly fine.

Reply:
Have change the "to int" casting in SingleSelectField + MultiSelectField to happen in getValueForValidation(), previously it was in setSubmittedValue().

This means value that stored internally in the FormField remains as a numeric string, rather than an int.  When it gets saved in DBVarchar there's not no validation error

DBInt will cast numeric strings to int, so things will continue work if we saved to a DBInt instead of a DBVarchar

Reply:
I've also updated this behaviour for NumericField by largely rewritting the cast() method.  Idea is that numeric strings are used for the most part, though cast to int/float for getValueForValidation()